### PR TITLE
[#152971091] Add documentation for Cerego's API

### PR DESCRIPTION
**Pull request by zaguios**
https://www.pivotaltracker.com/story/show/152971091

## Cerego's new documentation:

https://ceregousa.github.io/slate/

### CODE REVIEW

**Comment by johnstoecker** [comment]
Thats not your real token is it?

**Chris:** Lol, no, it's not my real token.

File: source/index.html.md [hunk 1]
```diff
@@ -19,221 +15,821 @@ search: true
...
+    -d '{"echo":"Test"}' 
+    -X POST 
+    -H "Content-Type: application/json"
+```
 
-This example API documentation page was created with [Slate](https://github.com/lord/slate). Feel free to edit it and use it as a base for your own API's documentation.
+> Please note that all other endpoints require authentication via a Bearer Token.
 
-# Authentication
+Welcome to the Cerego API! You can use our API to access Cerego endpoints, which can get information on about courses, assignments, users, and more.
```
Comment (kwstewar) [inline source/index.html.md]: Typo: on about

**Comment by albertchae** [comment]
Should we be publicizing stable and testing? Also is there a better way to get the token than the dev console? I thought Alex wrote a one off API for some partner, maybe we can clean that up for general use

**Comment by johnstoecker** [comment]
I usually use cerego.com/configuration to get the token

**Chris:** This is a much easier way to get the token, I'll change the documentation to use this method.

**Comment by albertchae** [comment]
LOL @johnstoecker hard lessons learned from ASU
*John:* there might have been some auth token exposure, but lets not just throw blame around willy nilly

File: source/index.html.md [hunk 2]
```diff
@@ -19,221 +15,821 @@ search: true
...
 
-```ruby
-require 'kittn'
+Cerego provides 3 different environments, 2 for testing and QA, as well as our primary production environment
 
-api = Kittn::APIClient.authorize!('meowmeowmeow')
-```
+`https://cerego.com/` - Production <br>
+`https://stable.cerego.com/` - Stable (Used for QA) <br>
+`https://testing.cerego.com/` - Testing (Used for heavy testing and unreleased features)
```
Comment (kwstewar) [inline source/index.html.md]: Don't know if we need to tell external people about this since it's pretty unstable.

**Comment by albertchae** [comment]
cerego.com/configuration looks good and TIL

File: source/index.html.md [hunk 3]
```diff
@@ -19,221 +15,821 @@ search: true
...
-Kittn uses API keys to allow access to the API. You can register a new Kittn API key at our [developer portal](http://example.com/developers).
+`cDPuiaEQrttyDpGU8a1aJtltFUDJY9W31Hi/K+SY9c2WuqHio3dBVtBjagLxyh6U` 
 
-Kittn expects for the API key to be included in all API requests to the server in a header that looks like the following:
+This is your Bearer token that you will need to use for API calls.
 
-`Authorization: meowmeowmeow`
+Cerego expects for Bearer token to be included in all API requests to the server in a header that looks like the following:
+
+`Authorization: Bearer cDPuiaEQrttyDpGU8a1aJtltFUDJY9W31Hi/K+SY9c2WuqHio3dBVtBjagLxyh6U`
```
Comment (albertchae) [inline source/index.html.md]: everywhere with a bearer token i'd do `<BEARER_TOKEN>` instead

File: source/index.html.md [hunk 4]
```diff
@@ -19,221 +15,821 @@ search: true
...
-```
+This endpoint retrieves all courses.
 
-```javascript
-const kittn = require('kittn');
+### HTTP Request
 
-let api = kittn.authorize('meowmeowmeow');
-let kittens = api.kittens.get();
+`GET https://cerego.com/api/v3/courses`
```
Comment (kwstewar) [inline source/index.html.md]: This gives me a 500, are certain parameters required?

File: source/index.html.md [hunk 5]
```diff
@@ -19,221 +15,821 @@ search: true
...
+
+### Course Object
+
+Parameter | type | Description
+--------- | --------- | -----------
+created_at | datetime | When the course was created
+name | string | The name of the course
+description | string | The description of the course
+slug | string | Unique name for the course
+users-count | integer | The amount of users in the course
```
Comment (kwstewar) [inline source/index.html.md]: Change amount to number in all of these fields that are counts.

File: source/index.html.md [hunk 6]
```diff
@@ -19,221 +15,821 @@ search: true
...
+filter[state] | `published` will display all courses that are in a published state <br> `unpublished` will display all courses that have not yet been published <br> `archived` will display all courses that have been archived
+page[number] | Courses are paginated, use this to choose which page you want
+page[size] | The default is `15` - How many courses are listed in each request/page
+tab | `published` will display all courses that are in a published state <br> `unpublished` will display all courses that have not yet been published <br> `archived` will display all courses that have been archived
+sort | Orders the results based on the param. <br>Example: `name` - sort in alphabetical order
+
+
+### Course Object
+
+Parameter | type | Description
```
Comment (kwstewar) [inline source/index.html.md]: Rename parameter to attribute.

File: source/index.html.md (hunk 2 above)
Comment (zaguios) [reply source/index.html.md]: I'm okay either way, would you prefer removing it?

File: source/index.html.md (hunk 4 above)
Comment (zaguios) [reply source/index.html.md]: I can query /courses with no params locally and get results.

File: source/index.html.md (hunk 2 above)
Comment (zaguios) [reply source/index.html.md]: I do kind of like the idea that people can use testing and stable to test the API before making their first calls that modify actual production data.

Comment (albertchae) [reply source/index.html.md]: I disagree. I don't think we want to be debugging stable/testing except for privileged partners like Cengage, and even them we don't want on testing. They should create their own testing data on prod to test their API with.

Comment (zaguios) [reply source/index.html.md]: Fair enough, I've removed all references to testing/stable

**Comment by darwin67** [comment]
how are you planning to publish the change?
are you going to manually push to the `gh-pages` branch every time after merging into `master`?

**Comment by zaguios** [comment]
To deploy all that's needed is to use the command: `./deploy.sh` It will be required to run this after merging changes that you want to appear.

File: source/index.html.md [hunk 7]
```diff
@@ -19,221 +15,807 @@ search: true
...
+Cerego utilizes the concept of bearer tokens to authenticate API calls. 
 
-> Make sure to replace `meowmeowmeow` with your API key.
+To easily acquire your bearer token please log into your [Cerego](https://cerego.com/) account (that has permission for the calls you wish to make) and then [click here.](https://cerego.com/configuration)
 
-Kittn uses API keys to allow access to the API. You can register a new Kittn API key at our [developer portal](http://example.com/developers).
+On the configuration page you will be able to get your Bearer token that you will need to use for API calls.
 
-Kittn expects for the API key to be included in all API requests to the server in a header that looks like the following:
+Cerego expects for Bearer token to be included in all API requests to the server in a header. The header should look like:
```
Comment (kwstewar) [inline source/index.html.md]: expects the

**Comment by darwin67** [comment]
cool.
also another random thought. 
would it be better to duplicate the repository instead of forking?
https://help.github.com/articles/duplicating-a-repository/

i think you would want to apply updates from upstream every once in a while.

**Comment by zaguios** [comment]
We could mirror it, the Slate documentation recommended just forking it. Do you think it's important enough to do?

**Comment by johnstoecker** [comment]
I think forking is the more common way of making a new repo -- we should still be able to apply upstream updates when needed

**Comment by albertchae** [comment]
I thought forking is usually with intent to merge stuff back to the original though? but probably doesn't really matter

**Comment by darwin67** [comment]
maybe. i personally would like to automate maintenance and updates as much as possible, so it's less of a headache to deal with. especially updating dependencies and stuff.
whichever way, create a different branch for our changes instead of merging it into `master`. keep master clean and easily synced with upstream, then rebase our changes on top when we need to do an update.

**Comment by albertchae** [comment]
So I see the pivotal ticket is only for v3/course and v3/course/users... I'm a bit wary about having a public API for all our endpoints because it means we can't make changes to our interface as easily. Not sure how much of a problem this will be in practice but just something to think about. Also it will require us to be way more careful about security, but that's a good thing anyway

**Comment by darwin67** [comment]
one reason for not doing a fork is, in case upstream repo got deleted for whatever random reason, this repo will be deleted too and your work will be gone for good.

**Comment by albertchae** [comment]
We could always restore from a local instance in that case

**Comment by zaguios** [comment]
@darwin67 I created a new stable branch that everything can be merged into to keep master synced with upstream.

@albertchae I think we should keep all of our endpoints documented regardless just to make it easier to see how the endpoints work internally. We don't necessarily need to expose it, but having it would still be nice. Maybe for now we can add a disclaimer that the API is not stable and backward incompatible changes may be made without warning?

I'll let you all make the final decision about whether or not mirroring the repo instead of forking it is necessary. I haven't maintained something for long enough to have a strong opinion about it.

**Comment by albertchae** [comment]
Documenting all our endpoints is good practice so I'm fine with that, but we should still carefully monitor who uses our endpoints via API. Even warning people things might not be backwards compatible still won't stop certain people from using them. Also we might have to think about how to rate limit API usage

**Comment by avolkovi** [comment]
@albertchae we can use rack-attack or rack-throttle for rate limiting, both are pretty easy to implement

**Pull request merged**